### PR TITLE
Add hardening options to CFLAGS by default if they are available.

### PR DESCRIPTION
**Pull request by Ferroin**
##### Summary

This adds checks to our configure script to enable a number of hardening features if they are available (and not overridden by the user), improving the baseline security of the Netdata Agent on many systems.

Currently adds:

- `-fstack-protector`/`-fstack-protector-strong`: This enables stack smashing protection, a form of buffer overflow protection designed to protect against exploits that overwrite parts of the stack. Support is largely ubiquitous and this is enabled by default in the compiler configuration on a vast majority of systems, so this change is unlikely to have a significant impact, but is being done anyway for the sake of correctness.
- `-fstack-clash-protection`: This enables a set of changes to how stack allocations are performed, making it more difficult to bypass the protections made by `-fstack-protector`. Support is relatively common, but this is not enabled by default on as many systems as `-fstack-protector` is so it is likely to have a practical impact for us to enable it.
- `-fcf-protection=full`: This enables Control Flow Integrity support, a hardware-assisted mechanism to avoid exploitation of ROP gadgets, which in turn protects against a whole category of exploits. Support is relatively new (only available since GCC 12 and Clang 9, and it requires build time options for those compilers for it to be available) and requires associated hardware support (which the compiler will check for before making use of this functionality). Overhead is comparatively low, but due to how new it is it’s not enabled on most platforms. Currently requires a recent 64-bit x86 CPU to gain any benefit, but other platforms may add similar functionality in the future.
- `-mbranch-protection=standard`: This enables hardware-assisted branch protection, which is conceptually similar to Control Flow Integrity. Much like `-fcf-protection`, this is relatively new and requires hardware support (but will be handled correctly by the compiler if such support is lacking). Currently requires a 64-bit ARM CPU to gain any benefit, but other platforms may add similar functionality in the future.
- `-D_FORTIFY_SOURCE=2`/`-D_FORTIFY_SOURCE=3`: These macros enable additional compile and runtime checking around functions that copy memory. They require both libc support (glibc fully supports them, some musl systems do as well but not all) and compiler support (a long list of compiler builtins is required), but supplying them when the libc support is lacking is safe. The checking provided by these options is generally low risk and low overhead, while providing a nontrivial improvement to security and also catching some questionable code at compile time.

##### Test Plan

Preliminary testing involves confirming that CI passes on this PR.

Full testing involves comparing performance of builds created with this PR to builds created using the master branch on the same systems and confirming that the performance overhead is acceptable (I expect it will be, but we still need to confirm this).

##### Additional Information

Relevant to: #15069 

Further things to consider for future investigation:

- `-z RELRO` and `-z NOW` linker flags.
- `-fvtable-verify` C++ compiler flag.

### CODE REVIEW

**Review by thiagoftsm — APPROVED** [review]
I tested this PR on:

- Slackware current
- FreeBSD 13.2
- Oracle 8.6
- Rocky Linux 8.8

And everything worked as expected, LGTM!

**Review by Dim-P — COMMENTED** [review]
All looks good while testing on:

- openSUSE Leap 15.4
- Debian 12
- Fedora 37

@stelfrag Do you think these changes require any benchmarking?

File: packaging/makeself/jobs/70-netdata-git.install.sh [hunk 1]
```diff
@@ -9,7 +9,7 @@ cd "${NETDATA_SOURCE_PATH}" || exit 1
 if [ "${NETDATA_BUILD_WITH_DEBUG}" -eq 0 ]; then
   export CFLAGS="-static -O2 -I/openssl-static/include -I/libnetfilter-acct-static/include/libnetfilter_acct -I/usr/include/libmnl -pipe"
 else
-  export CFLAGS="-static -O1 -pipe -ggdb -Wall -Wextra -Wformat-signedness -fstack-protector-all -D_FORTIFY_SOURCE=2 -DNETDATA_INTERNAL_CHECKS=1 -I/openssl-static/include -I/libnetfilter-acct-static/include/libnetfilter_acct -I/usr/include/libmnl"
+  export CFLAGS="-static -O1 -pipe -ggdb -Wall -Wextra -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1 -I/openssl-static/include -I/libnetfilter-acct-static/include/libnetfilter_acct -I/usr/include/libmnl"
```
[the remark this answers is not in the document]
Comment (Ferroin) [reply packaging/makeself/jobs/70-netdata-git.install.sh:12]: Excellent point, not sure how I missed these.

File: build_external/clean-install.Dockerfile [hunk 2]
```diff
@@ -26,8 +26,8 @@ RUN rm -rf autom4te.cache
 RUN rm -rf .git/
 RUN find . -type f >/opt/netdata/manifest
 
-RUN CFLAGS="-O1 -ggdb -Wall -Wextra -Wformat-signedness -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1\
-    -D_FORTIFY_SOURCE=2 -DNETDATA_VERIFY_LOCKS=1 ${EXTRA_CFLAGS}" ./netdata-installer.sh --disable-lto
+RUN CFLAGS="-O1 -ggdb -Wall -Wextra -Wformat-signedness -DNETDATA_INTERNAL_CHECKS=1\
```
Comment (thiagoftsm) [inline build_external/clean-install.Dockerfile:29]: @Ferroin we are having different `haldoint` warnings, please, take a look on them.

Comment (Dim-P) [reply build_external/clean-install.Dockerfile:29]: FWIW, the warnings are unrelated to this PR @thiagoftsm

**Pull request merged**